Stack/Opc.Ua.Core/Security/Certificates/CertificateFactory.cs

```diff
@@ -415,9 +444,326 @@ string password
+
+                    ISignatureFactory signatureFactory =
+                            new Asn1SignatureFactory(GetRSAHashAlgorithm(defaultHashSize), signingKey, random);
+
+                    X509V2CrlGenerator crlGen = new X509V2CrlGenerator();
+                    crlGen.SetIssuerDN(bcCertCA.IssuerDN);
+                    crlGen.SetThisUpdate(DateTime.UtcNow);
+                    crlGen.SetNextUpdate(DateTime.UtcNow.AddMonths(12));
+
+                    // merge all existing revocation list
+                    X509CrlParser parser = new X509CrlParser();
+                    foreach (X509CRL caCrl in certCACrl)
+                    {
+                        X509Crl crl = parser.ReadCrl(caCrl.RawData);
+                        crlGen.AddCrl(crl);
+                        var crlVersion = GetCrlNumber(crl);
+                        if (crlVersion.IntValue > crlSerialNumber.IntValue)
+                        {
+                            crlSerialNumber = crlVersion;
+                        }
+                    }
+
+                    if (certIsTheCA)
+                    {
+                        // add a dummy revoked cert
+                        crlGen.AddCrlEntry(BigInteger.One, DateTime.UtcNow, CrlReason.Superseded);
+                    }
+                    else
+                    {
+                        // add the revoked cert
+                        crlGen.AddCrlEntry(GetSerialNumber(certificate), DateTime.UtcNow, CrlReason.PrivilegeWithdrawn);
+                    }
+
+                    crlGen.AddExtension(X509Extensions.AuthorityKeyIdentifier,
+                                       false,
+                                       new AuthorityKeyIdentifierStructure(bcCertCA));
+
+                    // set new serial number
+                    crlSerialNumber = crlSerialNumber.Add(BigInteger.One);
+                    crlGen.AddExtension(X509Extensions.CrlNumber,
+                                       false,
+                                       new CrlNumber(crlSerialNumber));
+
+                    // generate updated CRL
+                    Org.BouncyCastle.X509.X509Crl updatedCrl = crlGen.Generate(signatureFactory);
+
+                    // add updated CRL to store
+                    X509CRL updatedCRL = new X509CRL(updatedCrl.GetEncoded());
+                    store.AddCRL(updatedCRL);
+
+                    // delete outdated CRLs from store
+                    foreach (X509CRL caCrl in certCACrl)
+                    {
+                        store.DeleteCRL(caCrl);
+                    }
+                }
+                store.Close();
+            }
+        }
+        catch (Exception e)
+        {
+            throw e;
+        }
+    }
+
+    /// <summary>
+    /// Creates a certificate signing request from an existing certificate.
+    /// </summary>
+    public static byte[] CreateSigningRequest(
```

Comment:
Can this method be used in ApplicationsNodeManager.cs from the GDS?

Reply:
is already used.